File: apple/src/Packages.m
/*
 * Copyright (c) 2021 Taner Sener
 *
 * This file is part of FFmpegKit.
 *
 * FFmpegKit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 *  You should have received a copy of the GNU Lesser General License
 *  along with FFmpegKit.  If not, see <http://www.gnu.org/licenses/>.
 */

#import "config.h"
#import "libavutil/ffversion.h"
#import "FFmpegKitConfig.h"
#import "Packages.h"

static NSMutableArray *supportedExternalLibraries;

@implementation Packages

+ (void)initialize {
    supportedExternalLibraries = [[NSMutableArray alloc] init];
    [supportedExternalLibraries addObject:@"dav1d"];
    [supportedExternalLibraries addObject:@"fontconfig"];
    [supportedExternalLibraries addObject:@"freetype"];
    [supportedExternalLibraries addObject:@"fribidi"];
    [supportedExternalLibraries addObject:@"gmp"];
    [supportedExternalLibraries addObject:@"gnutls"];
    [supportedExternalLibraries addObject:@"kvazaar"];
    [supportedExternalLibraries addObject:@"mp3lame"];
    [supportedExternalLibraries addObject:@"libaom"];
    [supportedExternalLibraries addObject:@"libass"];
    [supportedExternalLibraries addObject:@"iconv"];
    [supportedExternalLibraries addObject:@"libilbc"];
    [supportedExternalLibraries addObject:@"libtheora"];
    [supportedExternalLibraries addObject:@"libvidstab"];
    [supportedExternalLibraries addObject:@"libvorbis"];
    [supportedExternalLibraries addObject:@"libvpx"];
    [supportedExternalLibraries addObject:@"libwebp"];
    [supportedExternalLibraries addObject:@"libxml2"];
    [supportedExternalLibraries addObject:@"opencore-amr"];
    [supportedExternalLibraries addObject:@"openh264"];
    [supportedExternalLibraries addObject:@"opus"];
    [supportedExternalLibraries addObject:@"rubberband"];
    [supportedExternalLibraries addObject:@"sdl2"];
    [supportedExternalLibraries addObject:@"shine"];
    [supportedExternalLibraries addObject:@"snappy"];
    [supportedExternalLibraries addObject:@"soxr"];
    [supportedExternalLibraries addObject:@"speex"];
    [supportedExternalLibraries addObject:@"tesseract"];
    [supportedExternalLibraries addObject:@"twolame"];
    [supportedExternalLibraries addObject:@"x264"];
    [supportedExternalLibraries addObject:@"x265"];
    [supportedExternalLibraries addObject:@"xvid"];
}

+ (NSString*)getBuildConf {
    return [NSString stringWithUTF8String:FFMPEG_CONFIGURATION];
}

+ (NSString*)getPackageName {
    NSArray *enabledLibraryArray = [Packages getExternalLibraries];
    Boolean speex = [enabledLibraryArray containsObject:@"speex"];
    Boolean fribidi = [enabledLibraryArray containsObject:@"fribidi"];
    Boolean gnutls = [enabledLibraryArray containsObject:@"gnutls"];
    Boolean xvid = [enabledLibraryArray containsObject:@"xvid"];

    Boolean minGpl = false;
    Boolean https = false;
    Boolean httpsGpl = false;
    Boolean audio = false;
    Boolean video = false;
    Boolean full = false;
    Boolean fullGpl = false;

    if (speex && fribidi) {
        if (xvid) {
            fullGpl = true;
        } else {
            full = true;
        }
    } else if (speex) {
        audio = true;
    } else if (fribidi) {
        video = true;
    } else if (xvid) {
        if (gnutls) {
            httpsGpl = true;
        } else {
            minGpl = true;
        }
    } else {
        if (gnutls) {
            https = true;
        }
    }

    if (fullGpl) {
        if ([enabledLibraryArray containsObject:@"dav1d"] &&
            [enabledLibraryArray containsObject:@"fontconfig"] &&
            [enabledLibraryArray containsObject:@"freetype"] &&
            [enabledLibraryArray containsObject:@"fribidi"] &&
            [enabledLibraryArray containsObject:@"gmp"] &&
            [enabledLibraryArray containsObject:@"gnutls"] &&
            [enabledLibraryArray containsObject:@"kvazaar"] &&
            [enabledLibraryArray containsObject:@"mp3lame"] &&
            [enabledLibraryArray containsObject:@"libass"] &&
            [enabledLibraryArray containsObject:@"iconv"] &&
            [enabledLibraryArray containsObject:@"libilbc"] &&
            [enabledLibraryArray containsObject:@"libtheora"] &&
            [enabledLibraryArray containsObject:@"libvidstab"] &&
            [enabledLibraryArray containsObject:@"libvorbis"] &&
            [enabledLibraryArray containsObject:@"libvpx"] &&
            [enabledLibraryArray containsObject:@"libwebp"] &&
            [enabledLibraryArray containsObject:@"libxml2"] &&
            [enabledLibraryArray containsObject:@"opencore-amr"] &&
            [enabledLibraryArray containsObject:@"opus"] &&
            [enabledLibraryArray containsObject:@"shine"] &&
            [enabledLibraryArray containsObject:@"snappy"] &&
            [enabledLibraryArray containsObject:@"soxr"] &&
            [enabledLibraryArray containsObject:@"speex"] &&
            [enabledLibraryArray containsObject:@"twolame"] &&
            [enabledLibraryArray containsObject:@"x264"] &&
            [enabledLibraryArray containsObject:@"x265"] &&
            [enabledLibraryArray containsObject:@"xvid"]) {
            return @"full-gpl";
        } else {
            return @"custom";
        }
    }

    if (full) {
        if ([enabledLibraryArray containsObject:@"dav1d"] &&
            [enabledLibraryArray containsObject:@"fontconfig"] &&
            [enabledLibraryArray containsObject:@"freetype"] &&
            [enabledLibraryArray containsObject:@"fribidi"] &&
            [enabledLibraryArray containsObject:@"gmp"] &&
            [enabledLibraryArray containsObject:@"gnutls"] &&
            [enabledLibraryArray containsObject:@"kvazaar"] &&
            [enabledLibraryArray containsObject:@"mp3lame"] &&
            [enabledLibraryArray containsObject:@"libass"] &&
            [enabledLibraryArray containsObject:@"iconv"] &&
            [enabledLibraryArray containsObject:@"libilbc"] &&
            [enabledLibraryArray containsObject:@"libtheora"] &&
            [enabledLibraryArray containsObject:@"libvorbis"] &&
            [enabledLibraryArray containsObject:@"libvpx"] &&
            [enabledLibraryArray containsObject:@"libwebp"] &&
            [enabledLibraryArray containsObject:@"libxml2"] &&
            [enabledLibraryArray containsObject:@"opencore-amr"] &&
            [enabledLibraryArray containsObject:@"opus"] &&
            [enabledLibraryArray containsObject:@"shine"] &&
            [enabledLibraryArray containsObject:@"snappy"] &&
            [enabledLibraryArray containsObject:@"soxr"] &&
            [enabledLibraryArray containsObject:@"speex"] &&
            [enabledLibraryArray containsObject:@"twolame"]) {
            return @"full";
        } else {
            return @"custom";
        }
    }

    if (video) {
        if ([enabledLibraryArray containsObject:@"dav1d"] &&
            [enabledLibraryArray containsObject:@"fontconfig"] &&
            [enabledLibraryArray containsObject:@"freetype"] &&
            [enabledLibraryArray containsObject:@"fribidi"] &&
            [enabledLibraryArray containsObject:@"kvazaar"] &&
            [enabledLibraryArray containsObject:@"libass"] &&
            [enabledLibraryArray containsObject:@"iconv"] &&
            [enabledLibraryArray containsObject:@"libtheora"] &&
            [enabledLibraryArray containsObject:@"libvpx"] &&
            [enabledLibraryArray containsObject:@"libwebp"] &&
            [enabledLibraryArray containsObject:@"snappy"]) {
            return @"video";
        } else {
            return @"custom";
        }
    }

    if (audio) {
        if ([enabledLibraryArray containsObject:@"mp3lame"] &&
            [enabledLibraryArray containsObject:@"libilbc"] &&
            [enabledLibraryArray containsObject:@"libvorbis"] &&
            [enabledLibraryArray containsObject:@"opencore-amr"] &&
            [enabledLibraryArray containsObject:@"opus"] &&
            [enabledLibraryArray containsObject:@"shine"] &&
            [enabledLibraryArray containsObject:@"soxr"] &&
            [enabledLibraryArray containsObject:@"speex"] &&
            [enabledLibraryArray containsObject:@"twolame"]) {
            return @"audio";
        } else {
            return @"custom";
        }
    }

    if (httpsGpl) {
        if ([enabledLibraryArray containsObject:@"gmp"] &&
            [enabledLibraryArray containsObject:@"gnutls"] &&
            [enabledLibraryArray containsObject:@"libvidstab"] &&
            [enabledLibraryArray containsObject:@"x264"] &&
            [enabledLibraryArray containsObject:@"x265"] &&
            [enabledLibraryArray containsObject:@"xvid"]) {
            return @"https-gpl";
        } else {
            return @"custom";
        }
    }

    if (https) {
        if ([enabledLibraryArray containsObject:@"gmp"] &&
            [enabledLibraryArray containsObject:@"gnutls"]) {
            return @"https";
        } else {
            return @"custom";
        }
    }

    if (minGpl) {
        if ([enabledLibraryArray containsObject:@"libvidstab"] &&
            [enabledLibraryArray containsObject:@"x264"] &&
            [enabledLibraryArray containsObject:@"x265"] &&
            [enabledLibraryArray containsObject:@"xvid"]) {
            return @"min-gpl";
        } else {
            return @"custom";
        }
    }

    return @"min";
}

+ (NSArray*)getExternalLibraries {
    NSString *buildConfiguration = [Packages getBuildConf];
    NSMutableArray *enabledLibraryArray = [[NSMutableArray alloc] init];

    for (int i=0; i < [supportedExternalLibraries count]; i++) {
        NSString *supportedExternalLibrary = [supportedExternalLibraries objectAtIndex:i];

        NSString *libraryName1 = [NSString stringWithFormat:@"enable-%@", supportedExternalLibrary];
        NSString *libraryName2 = [NSString stringWithFormat:@"enable-lib%@", supportedExternalLibrary];

        if ([buildConfiguration rangeOfString:libraryName1].location != NSNotFound || [buildConfiguration rangeOfString:libraryName2].location != NSNotFound) {
            [enabledLibraryArray addObject:supportedExternalLibrary];
        }
    }

    [enabledLibraryArray sortUsingSelector:@selector(compare:)];

    return enabledLibraryArray;
}

@end

File: linux/src/Packages.cpp
/*
 * Copyright (c) 2022 Taner Sener
 *
 * This file is part of FFmpegKit.
 *
 * FFmpegKit is free software: you can redistribute it and/or modify
 * it under the terms of the GNU Lesser General License as published by
 * the Free Software Foundation, either version 3 of the License, or
 * (at your option) any later version.
 *
 * FFmpegKit is distributed in the hope that it will be useful,
 * but WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 * GNU Lesser General License for more details.
 *
 *  You should have received a copy of the GNU Lesser General License
 *  along with FFmpegKit.  If not, see <http://www.gnu.org/licenses/>.
 */

#include "Packages.h"
#include "config.h"
#include <memory>
#include <algorithm>

std::string ffmpegkit::Packages::getPackageName() {
    std::shared_ptr<std::set<std::string>> enabledLibrarySet = getExternalLibraries();
    #define contains_ext_lib(element) enabledLibrarySet->find(element) != enabledLibrarySet->end()
    bool speex = contains_ext_lib("speex");
    bool fribidi = contains_ext_lib("fribidi");
    bool gnutls = contains_ext_lib("gnutls");
    bool xvid = contains_ext_lib("xvid");

    bool min = false;
    bool minGpl = false;
    bool https = false;
    bool httpsGpl = false;
    bool audio = false;
    bool video = false;
    bool full = false;
    bool fullGpl = false;

    if (speex && fribidi) {
        if (xvid) {
            fullGpl = true;
        } else {
            full = true;
        }
    } else if (speex) {
        audio = true;
    } else if (fribidi) {
        video = true;
    } else if (xvid) {
        if (gnutls) {
            httpsGpl = true;
        } else {
            minGpl = true;
        }
    } else {
        if (gnutls) {
            https = true;
        } else {
            min = true;
        }
    }

    if (fullGpl) {
        if (contains_ext_lib("dav1d") &&
            contains_ext_lib("fontconfig") &&
            contains_ext_lib("freetype") &&
            contains_ext_lib("fribidi") &&
            contains_ext_lib("gmp") &&
            contains_ext_lib("gnutls") &&
            contains_ext_lib("kvazaar") &&
            contains_ext_lib("mp3lame") &&
            contains_ext_lib("libass") &&
            contains_ext_lib("iconv") &&
            contains_ext_lib("libilbc") &&
            contains_ext_lib("libtheora") &&
            contains_ext_lib("libvidstab") &&
            contains_ext_lib("libvorbis") &&
            contains_ext_lib("libvpx") &&
            contains_ext_lib("libwebp") &&
            contains_ext_lib("libxml2") &&
            contains_ext_lib("opencore-amr") &&
            contains_ext_lib("opus") &&
            contains_ext_lib("shine") &&
            contains_ext_lib("snappy") &&
            contains_ext_lib("soxr") &&
            contains_ext_lib("speex") &&
            contains_ext_lib("twolame") &&
            contains_ext_lib("x264") &&
            contains_ext_lib("x265") &&
            contains_ext_lib("xvid")) {
            return "full-gpl";
        } else {
            return "custom";
        }
    }

    if (full) {
        if (contains_ext_lib("dav1d") &&
            contains_ext_lib("fontconfig") &&
            contains_ext_lib("freetype") &&
            contains_ext_lib("fribidi") &&
            contains_ext_lib("gmp") &&
            contains_ext_lib("gnutls") &&
            contains_ext_lib("kvazaar") &&
            contains_ext_lib("mp3lame") &&
            contains_ext_lib("libass") &&
            contains_ext_lib("iconv") &&
            contains_ext_lib("libilbc") &&
            contains_ext_lib("libtheora") &&
            contains_ext_lib("libvorbis") &&
            contains_ext_lib("libvpx") &&
            contains_ext_lib("libwebp") &&
            contains_ext_lib("libxml2") &&
            contains_ext_lib("opencore-amr") &&
            contains_ext_lib("opus") &&
            contains_ext_lib("shine") &&
            contains_ext_lib("snappy") &&
            contains_ext_lib("soxr") &&
            contains_ext_lib("speex") &&
            contains_ext_lib("twolame")) {
            return "full";
        } else {
            return "custom";
        }
    }

    if (video) {
        if (contains_ext_lib("dav1d") &&
            contains_ext_lib("fontconfig") &&
            contains_ext_lib("freetype") &&
            contains_ext_lib("fribidi") &&
            contains_ext_lib("kvazaar") &&
            contains_ext_lib("libass") &&
            contains_ext_lib("iconv") &&
            contains_ext_lib("libtheora") &&
            contains_ext_lib("libvpx") &&
            contains_ext_lib("libwebp") &&
            contains_ext_lib("snappy")) {
            return "video";
        } else {
            return "custom";
        }
    }

    if (audio) {
        if (contains_ext_lib("mp3lame") &&
            contains_ext_lib("libilbc") &&
            contains_ext_lib("libvorbis") &&
            contains_ext_lib("opencore-amr") &&
            contains_ext_lib("opus") &&
            contains_ext_lib("shine") &&
            contains_ext_lib("soxr") &&
            contains_ext_lib("speex") &&
            contains_ext_lib("twolame")) {
            return "audio";
        } else {
            return "custom";
        }
    }

    if (httpsGpl) {
        if (contains_ext_lib("gmp") &&
            contains_ext_lib("gnutls") &&
            contains_ext_lib("libvidstab") &&
            contains_ext_lib("x264") &&
            contains_ext_lib("x265") &&
            contains_ext_lib("xvid")) {
            return "https-gpl";
        } else {
            return "custom";
        }
    }

    if (https) {
        if (contains_ext_lib("gmp") &&
            contains_ext_lib("gnutls")) {
            return "https";
        } else {
            return "custom";
        }
    }

    if (minGpl) {
        if (contains_ext_lib("libvidstab") &&
            contains_ext_lib("x264") &&
            contains_ext_lib("x265") &&
            contains_ext_lib("xvid")) {
            return "min-gpl";
        } else {
            return "custom";
        }
    }

    return "min";
}

std::shared_ptr<std::set<std::string>> ffmpegkit::Packages::getExternalLibraries() {
    const std::set<const char*> supportedExternalLibraries{
        "dav1d",
        "fontconfig",
        "freetype",
        "fribidi",
        "gmp",
        "gnutls",
        "kvazaar",
        "mp3lame",
        "libaom",
        "libass",
        "iconv",
        "libilbc",
        "libtheora",
        "libvidstab",
        "libvorbis",
        "libvpx",
        "libwebp",
        "libxml2",
        "opencore-amr",
        "openh264",
        "opus",
        "rubberband",
        "sdl2",
        "shine",
        "snappy",
        "soxr",
        "speex",
        "tesseract",
        "twolame",
        "x264",
        "x265",
        "xvid"};
    std::string buildConfiguration(FFMPEG_CONFIGURATION);
    char libraryName1[50];
    char libraryName2[50];
    std::shared_ptr<std::set<std::string>> enabledLibrarySet = std::make_shared<std::set<std::string>>();

    std::for_each(supportedExternalLibraries.cbegin(), supportedExternalLibraries.cend(), [&](const char* supportedExternalLibrary) {
        sprintf(libraryName1, "enable-%s", supportedExternalLibrary);
        sprintf(libraryName2, "enable-lib%s", supportedExternalLibrary);

        if (buildConfiguration.find(libraryName1) != std::string::npos || buildConfiguration.find(libraryName2) != std::string::npos) {
            enabledLibrarySet->insert(supportedExternalLibrary);
        }
    });

    return enabledLibrarySet;
}
